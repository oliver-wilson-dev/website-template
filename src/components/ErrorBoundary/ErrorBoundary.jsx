/* eslint-disable no-console */
import React from 'react';
import ErrorIcon from './assets/error.svg';
import styles from './ErrorBoundary.module.css';

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { error: null };
  }

  static getDerivedStateFromError(error) {
    return { error };
  }

  componentDidCatch(error, errorInfo) {
    // You can also log the error to an error reporting service
    console.log(error, errorInfo);
  }

  render() {
    const {
      state: {
        error
      },
      props: {
        children
      }
    } = this;

    return error ? (
      <div className={styles.errorBoundary}>
        <h1>Something went wrong</h1>
        <ErrorIcon className={styles.icon} />
      </div>
    ) : children;
  }
}

export default ErrorBoundary;
