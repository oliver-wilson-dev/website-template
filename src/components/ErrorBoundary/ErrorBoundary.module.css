.errorBoundary {
  position: fixed;
  width: 300px;
  height: 100px;
  top: 50%;
  left: 50%;
  margin-top: -50px;
  margin-left: -150px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.icon {
  width: 80px;
  height: 80px;
  overflow: visible;
  animation: pulse 1s ease-in-out infinite alternate;
}

@keyframes pulse {
  0% {
    transform: scale(0.9);
  }
  100% {
    transform: scale(1.1);
  }
}
