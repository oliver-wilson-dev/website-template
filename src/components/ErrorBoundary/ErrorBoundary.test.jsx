import React from 'react';
import { mount } from 'enzyme';
import ErrorBoundary from './ErrorBoundary';

jest.mock('./assets/error.svg', () => {
  const ErrorIcon = () => null;

  return ErrorIcon;
});

const ChildNoError = () => <h1>Default Child</h1>;

const render = ({ child: Child = ChildNoError } =
{ children: ChildNoError }) => mount(
  <ErrorBoundary>
    <Child />
  </ErrorBoundary>
);

const originalConsole = console;

describe('ErrorBoundary', () => {
  beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
  });

  afterAll(() => {
    console.log = originalConsole.log;
    console.error = originalConsole.log;
  });
  describe('when there are no errors thrown', () => {
    it('should render correctly', () => {
      expect(render()).toMatchSnapshot();
    });
  });

  describe('when there is an error thrown by a child', () => {
    const errMsg = 'oh no!';
    const ChildThrowsError = () => {
      throw new Error(errMsg);
      return <h1>Error!</h1>;
    };

    it('should render the error UI', () => {
      expect(render({ child: ChildThrowsError })).toMatchSnapshot();
    });

    it('should not render the children', () => {
      const children = render({ child: ChildThrowsError }).find(ChildThrowsError);

      expect(children.exists()).toBe(false);
    });
  });
});
