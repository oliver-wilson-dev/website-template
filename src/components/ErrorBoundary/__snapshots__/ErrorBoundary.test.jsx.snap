// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary when there are no errors thrown should render correctly 1`] = `
<ErrorBoundary>
  <ChildNoError>
    <h1>
      Default Child
    </h1>
  </ChildNoError>
</ErrorBoundary>
`;

exports[`ErrorBoundary when there is an error thrown by a child should render the error UI 1`] = `
<ErrorBoundary>
  <div
    className="errorBoundary"
  >
    <h1>
      Something went wrong
    </h1>
    <ErrorIcon
      className="icon"
    />
  </div>
</ErrorBoundary>
`;
